sddl/sddlSplitter.go

```diff
@@ -0,0 +1,403 @@
+							if processSID {
+								// We have to process the sid string now.
+								sidString := strings.TrimSuffix(strings.TrimPrefix(s[SIDStart:v], "SID("), ")")
+
+								tx, err := TranslateSID(strings.TrimSpace(sidString))
+
+								// It seems like we should probably still add the string if we error out.
+								// However, this just gets handled exactly like we're not processing the SID.
+								// When the next event happens, we just add everything to the string, including the original SID.
+								if err == nil {
+									workingString += "SID(" + tx + ")"
+									lastAddPoint = v
+								}
+							}
+						}
+					}
+				}
+
+				if workingString != "" {
+					if lastAddPoint != len(s) {
+						workingString += s[lastAddPoint:]
+					}
+
+					s = workingString
+				}
+
+				output += s
+			} else {
+				output += s
+			}
+		}
+
+		output += ")"
+	}
+
+	return strings.TrimSpace(output)
+}
+
+func (a *ACLList) String() string {
+	output := a.Flags
+
+	for _, v := range a.ACLEntries {
+		output += "("
+
+		for k, s := range v.Sections {
+			if k > 0 {
+				output += ";"
+			}
+
+			output += s
+		}
+
+		output += ")"
+	}
+
+	return strings.TrimSpace(output)
+}
+
+func (s *SDDLString) String() string {
+	output := ""
+
+	if s.OwnerSID != "" {
+		output += "O:" + s.OwnerSID
+	}
+
+	if s.GroupSID != "" {
+		output += "G:" + s.GroupSID
+	}
+
+	if s.DACL.Flags != "" || len(s.DACL.ACLEntries) != 0 {
+		output += "D:" + s.DACL.String()
+	}
+
+	if s.SACL.Flags != "" || len(s.SACL.ACLEntries) != 0 {
+		output += "S:" + s.SACL.String()
+	}
+
+	return output
+}
+
+// place an element onto the current ACL
+func (s *SDDLString) putACLElement(element string, aclType rune) error {
+	var aclEntries *[]ACLEntry
+	switch aclType {
+	case 'D':
+		aclEntries = &s.DACL.ACLEntries
+	case 'S':
+		aclEntries = &s.SACL.ACLEntries
+	default:
+		return fmt.Errorf("%s ACL type invalid", string(aclType))
+	}
+
+	aclEntriesLength := len(*aclEntries)
+	if aclEntriesLength == 0 {
+		return errors.New("ACL Entries too short")
+	}
+
+	entry := (*aclEntries)[aclEntriesLength-1]
+	entry.Sections = append(entry.Sections, element)
+	(*aclEntries)[aclEntriesLength-1] = entry
+	return nil
+}
+
+// create a new ACL
+func (s *SDDLString) startACL(aclType rune) error {
+	var aclEntries *[]ACLEntry
+	switch aclType {
+	case 'D':
+		aclEntries = &s.DACL.ACLEntries
+	case 'S':
+		aclEntries = &s.SACL.ACLEntries
+	default:
+		return fmt.Errorf("%s ACL type invalid", string(aclType))
+	}
+
+	*aclEntries = append(*aclEntries, ACLEntry{Sections: make([]string, 0)})
+
+	return nil
+}
+
+func (s *SDDLString) setACLFlags(flags string, aclType rune) error {
+	var aclFlags *string
+	switch aclType {
+	case 'D':
+		aclFlags = &s.DACL.Flags
+	case 'S':
+		aclFlags = &s.SACL.Flags
+	default:
+		return fmt.Errorf("%s ACL type invalid", string(aclType))
+	}
+
+	*aclFlags = strings.TrimSpace(flags)
+
+	return nil
+}
+
+func ParseSDDL(input string) (sddl SDDLString, err error) {
```

Comment:
It would be great to have a big (ish) set of sample SDDLs to run through this in unit tests.  Because its complicated enough that something like that seems justified. 

Given that we don't _have_ such a set, how about we instead add a runtime-sanity check? That will run on everything that we ever parse.  To implement, we'd write code something like this

```
var sddlWhitespaceRegex = regexp.MustCompile(`[\x09-\x0D ]`)  // as per SDDL grammar

// sanityCheckSddlParse acts as a double check that we parsed the SDDL correctly,
// by seeing that we can round trip the result back to what we started with (minus any whitespace)
func sanityCheckSddlParse(originalInput string, parsed SDDLString) bool {
	return sddlWhitespaceRegex.ReplaceAllString(originalInput, "") ==
		sddlWhitespaceRegex.ReplaceAllString(parsed.String(), "")
}
```

and then call that at the very end of ParseSDDL like this

```
	if err == nil {
		ok := sanityCheckSddlParse(input, sddl)
		if !ok {
			return sddl, errors.New("SDDL parsing failed sanity check")
		}
	}
```

What do you think?

Reply:
I don't think the sanity check is a bad idea.

This _in theory_ should reproduce the exact same output, even without cleaning spaces

Reply:
_actually_, that's not feasible for SDDLString.PortableString.

The output would never be the same unless all SDDLs were already portable.